modules/basicfuncs/numeric-funcs.c

```diff
@@ -130,3 +132,156 @@ tf_num_mod(LogMessage *msg, gint argc, GString *argv[], GString *result)
+
+  return (GString *) g_ptr_array_index(args->bufs, 0);
+}
+
+static gboolean
+_tf_num_parse_arg_with_message(const TFSimpleFuncState *state,
+                               LogMessage *message,
+                               const LogTemplateInvokeArgs *args,
+                               gint64 *number)
+{
+  GString *formatted_template_result = _make_gstring_buffer(args);
+  gint on_error = args->opts->on_error;
+
+  log_template_format(state->argv[0], message, args->opts, args->tz,
+    args->seq_num, args->context_id, formatted_template_result);
+
+  if (!parse_number_with_suffix(formatted_template_result->str, number))
+    {
+      if (!(on_error & ON_ERROR_SILENT))
+        msg_error("Parsing failed, template function's argument is not a number",
+                  evt_tag_str("arg", formatted_template_result->str));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+tf_num_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent,
+               gint argc, gchar *argv[], GError **error)
+{
+  g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+  if (argc != 2)
+    {
+      g_set_error(error, LOG_TEMPLATE_ERROR, LOG_TEMPLATE_ERROR_COMPILE,
+        "Template function requires only one argument");
+      return FALSE;
+    }
+
+  return tf_simple_func_prepare(self, s, parent, argc, argv, error);
+}
+
+static void
+tf_num_sum_call(LogTemplateFunction *self, gpointer s,
+                const LogTemplateInvokeArgs *args, GString *result)
+{
+  TFSimpleFuncState *state = (TFSimpleFuncState *) s;
+  gint64 number;
+  gint64 sum = 0;
+
+  gint message_index;
+  for (message_index = 0; message_index < args->num_messages; ++message_index)
+    {
+      LogMessage *message = args->messages[message_index];
+      if (_tf_num_parse_arg_with_message(state, message, args, &number))
+        sum += number;
+    }
+
+  format_int64_padded(result, 0, ' ', 10, sum);
+}
+
+TEMPLATE_FUNCTION(TFSimpleFuncState, tf_num_sum,
+                  tf_num_prepare, NULL, tf_num_sum_call,
+                  tf_simple_func_free_state, NULL);
+
+static gboolean
+_tf_num_get_first_valid_arg(const TFSimpleFuncState *state,
+                            const LogTemplateInvokeArgs *args,
+                            gint *message_index, gint64 *number)
+{
+  for (*message_index = 0; *message_index < args->num_messages; (*message_index)++)
+    {
+      LogMessage *message = args->messages[*message_index];
+
+      if (_tf_num_parse_arg_with_message(state, message, args, number))
+        return TRUE;
+    }
+
+  return FALSE;
+}
+
+static gboolean
+_tf_num_extremum(const TFSimpleFuncState *state,
+                 const LogTemplateInvokeArgs *args,
+                 extremum_func extremum_compare, gint64 *extremum)
+{
+  gint64 number;
+
+  gint message_index;
+  if (!_tf_num_get_first_valid_arg(state, args, &message_index, extremum))
+    return FALSE;
+
+  for (; message_index < args->num_messages; ++message_index)
```

Comment:
We made a small refactor on this part with @bkil-syslogng to reduce the redundancies of iterating through the message context.